### PR TITLE
Added a function to create Client.Request in ClientEnv

### PR DESCRIPTION
**Pull request by symbiont-eric-torreborre**
Hi, I am proposing this PR to fix an issue I am currently facing but you might propose a better solution. Basically I need to set "dynamic timeouts" on HTTP requests based on previous attempts (I try a query, if it times out, I try with a bigger timeout).

Unfortunately at the moment when I get to execute a `ClientM` action there is nothing I can do to intercept the code in `performRequest` which is transforming a `Servant.Client.Core.Request` into a `Network.HTTP.Client.Request` (that function is requestToClientRequest). This PR adds such a function into the `ClientEnv` so that I can override it with some fine-tuning behaviour for changing the sent `Network.HTTP.Client.Requests`.

Some other options would be to more massively use `Free ClientF` in our code (instead of `ClientM`) and rely on a specific interpreter. The drawback would be that we would have to embed all the code in `performRequest` (in theory, because in our specific use case we don't rely on cookies so this would be less code).

What do you think? Do you see better alternatives?

### CODE REVIEW

**Comment by symbiont-eric-torreborre** [comment]
Note: I had to add some dependencies updates to the `stack.yaml` file other wise I would get errors like
```
Error: While constructing the build plan, the following exceptions were encountered:

In the dependencies for servant-0.16:
    http-api-data-0.3.8.1 from stack configuration does not match >=0.4.1 && <0.4.2  (latest matching version is 0.4.1)
needed due to servant-client-0.16 -> servant-0.16

Some different approaches to resolving this:

  * Set 'allow-newer: true' in /Users/etorreborre/.stack/config.yaml to ignore all version constraints and build anyway.

  * Recommended action: try adding the following to your extra-deps in /Users/etorreborre/projects/haskell/servant/stack.yaml:

- http-api-data-0.4.1@sha256:7a3d3782859c7cac74b82cd41ad8ae0ccd6d6ca5279a88b77d4421de71920a4a,4039

Plan construction failed.
```
Anyone knows why?

**Comment by alpmestan** [comment]
Hello! Sorry for the delay.

I'm personally not opposed to this change, especially given that in the absolute majority of situations people won't have to change their code. @phadej Objections?

File: servant-client/src/Servant/Client/Internal/HttpClient.hs [hunk 1]
```diff
@@ -73,11 +73,12 @@ data ClientEnv
   { manager :: Client.Manager
   , baseUrl :: BaseUrl
   , cookieJar :: Maybe (TVar Client.CookieJar)
+  , makeClientRequest :: BaseUrl -> Request -> Client.Request
```
Comment (phadej) [inline servant-client/src/Servant/Client/Internal/HttpClient.hs]: - this one needs haddocks
    - at leas mentioning that `cookieJar` is applied to the request created
- also we'll need to export default value from non-`Internal` module, as probably many implementations will call it either mangling servant's or http-client's `Request`. I propose `defaultMakeClientRequest` (it also needs haddocks, explaining what it does).

Comment (symbiont-eric-torreborre) [reply servant-client/src/Servant/Client/Internal/HttpClient.hs]: Thanks for the comment I will do that tomorrow.

**Comment by symbiont-eric-torreborre** [comment]
@phadej I amended my commit with:
 
 - some documentation on `ClientEnv` 
 - a similar change on the `ClientEnv` for streaming requests for consistency
 - a test where I show that it is possible to use the `ClientEnv` to modify the headers before executing a request

File: servant-client/src/Servant/Client/Internal/HttpClient.hs [hunk 2]
```diff
@@ -68,16 +68,23 @@ import qualified Servant.Types.SourceT                    as S
 import qualified Network.HTTP.Client         as Client
 
 -- | The environment in which a request is run.
+--   The baseUrl and makeClientRequest function are used to create a `http-client` request.
```
Comment (phadej) [inline servant-client/src/Servant/Client/Internal/HttpClient.hs]: Note, haddocks aren't markdown. 

```
The 'baseUrl' and 'makeClientRequest' function are used to create a @http-clietn@ request.

...

on a @servant@

```

In short: use ticks, no backticks. `@` for "code-like".

File: servant-client/test/Servant/SuccessSpec.hs [hunk 3]
```diff
@@ -125,16 +126,26 @@ successSpec = beforeAll (startWaiApp server) $ afterAll endWaiApp $ do
...
-      _ <- runClientM (getRedirectWithCookie HTTP.methodGet) (ClientEnv mgr baseUrl (Just cj))
+      _ <- runClientM (getRedirectWithCookie HTTP.methodGet) (ClientEnv mgr baseUrl (Just cj) defaultMakeClientRequest)
       cookie <- listToMaybe . C.destroyCookieJar <$> atomically (readTVar cj)
       C.cookie_name <$> cookie `shouldBe` Just "testcookie"
       C.cookie_value <$> cookie `shouldBe` Just "test"
 
+    it "Can modify the outgoing Request using the ClientEnv" $ \(_, baseUrl) -> do
+      mgr <- C.newManager C.defaultManagerSettings
+      let createClientRequest url r = (defaultMakeClientRequest url r) { C.requestHeaders = [("X-Added-Header", "XXX")] }
+      let clientEnv = ClientEnv mgr baseUrl Nothing createClientRequest
```
Comment (phadej) [inline servant-client/test/Servant/SuccessSpec.hs]:
```haskell
(mkClientEnv mg baseUrl)} { makeClientRequest = createClientRequest }
```

File: servant-client/src/Servant/Client/Internal/HttpClient.hs (hunk 2 above)
Comment (symbiont-eric-torreborre) [reply servant-client/src/Servant/Client/Internal/HttpClient.hs]: Thanks, I fixed the comments in a fixup commit.

File: servant-client/src/Servant/Client/Internal/HttpClient.hs [hunk 4]
```diff
@@ -68,16 +68,23 @@ import qualified Servant.Types.SourceT                    as S
...
+--   Finally the request is executed with the manager.
+--   The 'makeClientRequest' function can be used to modify the request to execute and set values which
+--   are not specified on a `servant` Request like the `responseTimeout` or the `redirectCount`
 data ClientEnv
   = ClientEnv
   { manager :: Client.Manager
   , baseUrl :: BaseUrl
   , cookieJar :: Maybe (TVar Client.CookieJar)
+  , makeClientRequest :: BaseUrl -> Request -> Client.Request
+  -- ^ this function can be used to customize the creation of `http-client` requests from `servant` requests. Default value: 'defaultMakeClientRequest'
```
Comment (phadej) [inline servant-client/src/Servant/Client/Internal/HttpClient.hs]: add also

`makeClientRequest` exists to allow overriding operational semantics e.g. `responseTimeout` *per request*.
If you need global modifications, you should use [`managerModifyRequest`](https://hackage.haskell.org/package/http-client-0.6.4/docs/Network-HTTP-Client.html#v:managerModifyRequest)

File: servant-client/test/Servant/SuccessSpec.hs [hunk 5]
```diff
@@ -125,16 +126,26 @@ successSpec = beforeAll (startWaiApp server) $ afterAll endWaiApp $ do
...
       cj <- atomically . newTVar $ C.createCookieJar []
-      _ <- runClientM (getRedirectWithCookie HTTP.methodGet) (ClientEnv mgr baseUrl (Just cj))
+      _ <- runClientM (getRedirectWithCookie HTTP.methodGet) (ClientEnv mgr baseUrl (Just cj) defaultMakeClientRequest)
       cookie <- listToMaybe . C.destroyCookieJar <$> atomically (readTVar cj)
       C.cookie_name <$> cookie `shouldBe` Just "testcookie"
       C.cookie_value <$> cookie `shouldBe` Just "test"
 
+    it "Can modify the outgoing Request using the ClientEnv" $ \(_, baseUrl) -> do
+      mgr <- C.newManager C.defaultManagerSettings
+      let createClientRequest url r = (defaultMakeClientRequest url r) { C.requestHeaders = [("X-Added-Header", "XXX")] }
```
Comment (phadej) [inline servant-client/test/Servant/SuccessSpec.hs]: Add a comment:

In proper situation, extra headers should probably visible in API type. However, testing for response timeout is difficult, so we test with something which is easy to observe.

File: servant-client/src/Servant/Client/Internal/HttpClient.hs (hunk 4 above)
Comment (phadej) [reply servant-client/src/Servant/Client/Internal/HttpClient.hs]: .. and the note that possible `cookieJar` is applied after.

**Comment by symbiont-eric-torreborre** [comment]
@phadej I added better comments

**Comment by phadej** [comment]
Good. Let me wait for CI, I'll merge them.

Thanks for your contribution.

**Comment by symbiont-eric-torreborre** [comment]
Thanks a lot for reviewing and accepting that change!

**Comment by symbiont-eric-torreborre** [comment]
@phadej I didn't see the docs indeed. The build is passing now.

**Comment by phadej** [comment]
Squashed in https://github.com/haskell-servant/servant/pull/1255

**Pull request closed without merging**